### PR TITLE
feat: deleted the pods that are not unmanaged by Cilium

### LINKED ISSUE #21594: Unmanaged pods with preemptible nodes on GKE
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

Our organisation is using GKE with [preemptible](https://cloud.google.com/compute/docs/instances/preemptible) node pools. The preemptible VMs used in those node pools aren't _always available_ so we see some VMs being stopped and recreated from time to time. We use these VMs to reduce cost and with proper taints, tolerations & node affinity configuration we're able to run non-sensitive workloads on these node pools while running other more sensitive workloads on proper VMs. 

When using Cilium with preemptible VMs in GKE, we've noticed that when a preemptible VM linked to a node is deleted, the underlying compute instance in GCP gets recreated but the k8s node _remains the same_. Because the k8s node isn’t recreated the node pool taint doesn’t get reapplied and we end up in situations where pods are potentially unmanaged if cilium isn't up and ready when other pods get created. Ultimately in these situations we then see legitimate traffic being blocked.

I wonder if somehow the `cilium-operator` could help in this situation: from what I can see the operator is already responsible for watching cilium-agent pods and removing the node taint once cilium is ready. Would it somehow be possible to taint back a node when the operator sees the cilium-agent not running/modified or deleted on a node?

### Cilium Version

Cilium v1.12.0

### Kernel Version

COS-5.10.133

### Kubernetes Version

v1.22.12-gke.1200

### Sysdump

_No response_

### Relevant log output

_No response_

### Anything else?

I did pop a message in the Cilium kubernetes slack channel too: https://cilium.slack.com/archives/C53TG4J4R/p1664808717519899  

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by lvyanru8200**
Signed-off-by: yanru.lv <yanru.lv@daocloud.io>

Please ensure your pull request adheres to the following guidelines:

- [ ] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [ ] All code is covered by unit and/or runtime tests where feasible.
- [ ] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [ ] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [ ] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [ ] Provide a title or release-note blurb suitable for the release notes.
- [ ] Thanks for contributing!

Fixes: https://github.com/cilium/cilium/issues/21594

### CODE REVIEW

**Comment by lvyanru8200** [comment]
cc @christarazi

**Comment by lvyanru8200** [comment]
cc @aanm

**Review by christarazi — CHANGES_REQUESTED** [review]
Thanks for the PR. Could you add the justification to the commit msg? Also, have you tested that this works?

**Comment by christarazi** [comment]
@lvyanru8200 Ping

**Review by joestringer — CHANGES_REQUESTED** [review]
I don't think we can just set this behaviour by default, especially since if Cilium is not correctly installed this may just cause all pods in the cluster to be inexplicably deleted constantly. At least if it's restricted to kube-dns then the impact is somewhat lower / more restricted and it gives a stronger hint that this is the specific issue.

I would suggest that this be exposed as a commandline argument like `--pod-restart-selector` which accepts a list of labels. It should have a default setting of `k8s-app=kube-dns`, then if _you_ want to allow all pods to be recreated, you would pass `--pod-restart-selector=""`.

**Comment by lvyanru8200** [comment]
> 

pong

**Comment by lvyanru8200** [comment]
> @lvyanru8200您是否提交了建议的更改@joestringer？

Wait a minute, before merging the commit

**Comment by thejosephstevens** [comment]
Are there any maintainers or @lvyanru8200 who could help recover this commit? I'm running into a similar issue as is described [here](https://github.com/cilium/cilium/issues/21594), and am trying to figure out the best way to address it. In the past we've deployed custom control loops to just diff between pods and CEP and delete any pods on the diff, but we had some weirdness in the results (pods getting deleted when it wasn't clear whether they should be), and I think it would make more sense for this to be an in-tree feature for Cilium.

We (I or someone on my team), could potentially take a stab at this, I think what we would be looking for is for the operator to monitor pods, and if they should be managed but aren't (similar to what helm does), it would kill them. Perfectly fine with it being toggleable behavior so this doesn't catch anyone by surprise.

tl;dr: Right now when a node restarts in GKE it doesn't look like the taint lifecycle is working as expected, so pods are becoming unmanaged and I'm trying to find a path to resolution.
